### ISSUE TITLE
docs:disk_interface::[async_move_storage()]

### ISSUE THREAD

**Issue by cheater** [arvidn/libtorrent#6725]
Documentation under heading "disk_interface::[async_move_storage()]" could be improved:

The documentation does not specify whether sparse files are inflated upon being moved to a different file system.

It's a pretty big problem for the following reasons, and it comes as a combination of things that aren't necessarily due to libtorrent, but it's the situation that libtorrent finds itself in and so it has to live with it.

First of all, for many residential users, seeding more than a handful of torrents will essentially ddos their router, as it just can't handle the network pressure associated with outgoing connections necessary to seed a torrent correctly. so you want to limit your seeding to maybe say 10-20 torrents.

Second of all, if you want to maximize your utility, you will be seeding collection torrents. For example, let's say there is a free (as in freedom) tv series that is released both on an episodic basis (one torrent per episode), and everything in one torrent (4 sesons / 40 episodes in one torrent).
You want to be seeding the second one, because the internet drop outs will actually be /per torrent/ not /per megabytes shared/ so the more megabytes you share the better.
Now the next aspect of this is that most people using torrents a lot will have a temporary disk and then completed torrents are moved to another disk that only stores stuff and never has downloads going. Not doing things this way (i.e. you have both temp/ and download/ on the same FS) results in pathological file systems that slow down file IO by orders of magnitude.
The last aspect of this is that many people downloading collection torrents will not download everything. Let's say a musician released 500 of their music videos in one torrent and you only want, say, 20 of them. They will almost certainly not be contiguous. That means that for every file you want to download, the last chunk of the file before that will also be downloaded. In temp/, a sparse file will be created for that previous file that will only take up as much disk space as that last chunk needs. However, when this file is moved, the file will be inflated.

This means that for a torrent where you download 20 files, you will be essentially using the amount of disk space that 40 files take up.

The current behavior I'm seeing from clients using libtorrent is that sparse files are not being handled specially, which means that this space doubling happens.

**Comment by arvidn** [comment]
are you experiencing this on windows or linux?
I agree that the documentation should clarify this, and I also think the implementation should go to greater length at trying to preserve sparse regions. Currently it uses system functions `copyfile()` on MacOS and `CopyFileW()` on windows. It's not clear from their documentation whether they preserve sparse regions or not. But on other systems files are copied by simply reading blocks and writing them to the new file, so definitely not preserving sparse regions.

**Comment by cheater** [comment]
My experience is that sparse regions are not preserved on any platforms.

In addition to this, the files being sparse creates an issue when the user wants to move the files again, this time by hand. The files will be inflated by the file manager. Relying on sparse files is a bad idea. Files should be created as parts.

In addition to this, it is unclear why unwanted files must be present in the first place. Of course, torrent health is improved if a block is fully preserved so that it can be shared. But it should not come first over the preference of the user as to which files to keep and which files not to keep. If the unwanted partial files (sparse or not) get deleted, the torrent then registers as incomplete (usually around 99%) and so the files usually get all moved back to the temp FS, and then the 1% downloaded, and the unwanted files put back in place. The only way to stop this from happening is to pause the torrent which also stops it from seeding. I think due to this chain of events this behavior creates LESS availability than simply deleting the unwanted files.

**Comment by cheater** [comment]
I also wanted to mention that some thought needs to be added to the common situation where files already exist on disk and they have been inflated beyond the sparse form, and how to make them take up less space again.

**Comment by arvidn** [comment]
> In addition to this, the files being sparse creates an issue when the user wants to move the files again, this time by hand. The files will be inflated by the file manager. Relying on sparse files is a bad idea. Files should be created as parts.

How would that work? What do you do with the parts once you're done?

> In addition to this, it is unclear why unwanted files must be present in the first place.

It's easy to underestimate torrent health impact. Specifically, the health of the piece that partly overlap a file that virtually nobody wants to download, but also overlaps a file that virtually everybody wants to download. That piece could disappear from the swarm really easily if downloaders of it did not preserve it.

**Comment by arvidn** [comment]
> I also wanted to mention that some thought needs to be added to the common situation where files already exist on disk and they have been inflated beyond the sparse form, and how to make them take up less space again.

What's the use case for this? is it common to just want to download certain portions of a file, but not others, and then be done? I would have expected sparse files to primarily be an intermediate state while downloading. libtorrent will clear the sparse flag on windows once a file is completely downloaded.

**Comment by cheater** [comment]
A typical example is a torrent that contains the complete output of a performer. Let's say all albums of a musician, or say all episodes of a TV series. You might only want some of the albums, or some of the episodes, etc. Sometimes these are 500 GB to 1TB heavy collections and you will pick and choose what you want.

Some people will only download what they want. Some others will download everything. But people who realize their disk space is being wasted will inevitably not be seeding at all, so that's a net loss.

I don't know about libtorrent version. I'm using whatever qbittorrent is using.

**Comment by cheater** [comment]
> > Files should be created as parts.
> 
> How would that work? What do you do with the parts once you're done?

Sorry, I don't know why I wrote "Files should be created as parts". I think I lost my focus there.

**Comment by cheater** [comment]
What I meant there was that downloads should be created as files, not as parts. So if all your selected files are already downloaded, the % completion indicated should be 100%. There should be a separate completion percentage based on whether you want full chunks or not, and torrent clients should default to "don't want full chunks", with an option called something like "download whole chunks".

**Comment by arvidn** [comment]
I may be misunderstanding, it sounds like you're describing a UX issue. I believe libtorrent provides accurate counters to present whatever you're looking for. But if you just want per-file completion, that's available directly in the API.

**Comment by cheater** [comment]
Right, but there doesn't seem to be a way not download full chunks, i.e. not download files you didn't elect to download?

**Comment by arvidn** [comment]
> Right, but there doesn't seem to be a way not to download full chunks, i.e. not download files you didn't elect to download?

Right. downloaded bytes can only be validated at the piece level. If you download less than a piece, you can't validate those bytes.

**Comment by cheater** [comment]
True, but once the piece has been validated, the unneeded bytes can be tossed away.

As I understand extensions to the BitTorrent protocol are possible, so it might well be possible to offer partial pieces.

**Comment by arvidn** [comment]
> True, but once the piece has been validated, the unneeded bytes can be tossed away.

I think it's easy to underestimate the risk to the swarm if everybody implemented that behavior. If there are files that most people choose not to download, you may have a hard time finding that last piece of the file you *do* want to download, since  very few people has it. I believe this is worse that the normal issue with peers not picking certain pieces, because there's likely to be a strong bias towards certain files being de-selected.

It's not clear to me why this matters so much, it's quite rare these overlapping pieces amount to very much data anyway. You would need a very large piece size, and very many (small) files, and you select every other.

**Comment by arvidn** [comment]
> It happens in every torrent where files are large.

only if you're using a 3 years old copy of libtorrent, or older. It sounds like you would benefit from upgrading. Starting with libtorrent 1.2, pieces in files with priority 0 are stored in a separate partfile.

**Comment by cheater** [comment]
hmm... what's missing? qbittorrent uses libtorrent 1.2.14.0, and sparse files are used.

**Comment by cheater** [comment]
hmm... what's missing? qbittorrent uses libtorrent 1.2.14.0, and sparse files are used.

**Comment by arvidn** [comment]
do you set the file priority *after* those unwanted files have been created by chance?

**Comment by cheater** [comment]
possibly, yeah.

**Comment by cheater** [comment]
no, i take it back, i have had situations where the unwanted files were created even if i set the priority to "don't download" when first adding the torrent.

**Comment by arvidn** [comment]
ok. There is a missing feature to move data into a partfile.

https://github.com/arvidn/libtorrent/blob/RC_1_2/src/storage.cpp#L165

**Comment by cheater** [comment]
hmm, i see.

what if the file has had data downloaded into it, then i delete it from the disk, recheck the torrent, and restart it to go to 100%? does the rest of the part get put in the part file, too?

**Comment by arvidn** [comment]
in that case, chances are that you're missing one piece by the time you're done checking. If the unwanted file has priority 0, then the missing piece will be downloaded and the part of it overlapping the unwanted file should go into the partfile.

I'm not 100% sure about that. There is some logic for backwards compatibility where files that exist on disk when the torrent is added is remembered to not use the part-file. So it's possible this state isn't reconsidered by a re-check. But a restart of the client, or removing and re-adding the torrent definitely would.

**Comment by cheater** [comment]
this would be useful to have in the docs too :)

**Comment by arvidn** [comment]
https://github.com/arvidn/libtorrent/pull/6732

**Comment by cheater** [comment]
Thank you!

A few questions seem to remain.

1. in sparse files that we elect not to download, can downloaded parts be deleted? what should the UX be? if bit torrent stores gigabytes in files I ultimately don't want, that is annoying. But if I mis-click and uncheck a file that I've been downloading for months at 1 KB/s, and that file is instantly deleted without me being asked if I'm sure of this, then that would be just as intensely annoying.
2. can files that should be sparse, but aren't, be properly deflated? For example, a file that has 20% of its parts downloaded, but it's unchecked ("do not download"), and it has been moved, so it takes up 100% disk space. Each such file could easily free several gigabytes of space on the disk just by converting the on-disk zeros to holes (using sparse files).
3. unwanted sparse files being converted to parts files, possibly in tandem with what's going on in 2.

**Comment by arvidn** [comment]
>     1. in sparse files that we elect not to download, can downloaded parts be deleted? what should the UX be? if bit torrent stores gigabytes in files I ultimately don't want, that is annoying. But if I mis-click and uncheck a file that I've been downloading for months at 1 KB/s, and that file is instantly deleted without me being asked if I'm sure of this, then that would be just as intensely annoying.

The main reason those "gigabytes" are there is because you move the file to a different volume, and libtorrent's `copy_file()` isn't sophisticated enough to preserve sparse regions, right? So, fixing that would be an option I imagine.

Regarding moving data from a file (as specified in the .torrent) into a partfile when setting its priority to 0 is possible, but it's a bit complicated with libtorrent's current layering of responsibility. The disk subsystem doesn't track which parts of files are sparse and not, so when it's asked to change the priority of a file, It doesn't know which pieces of that file would need to be moved into the sparse file. It could guess by looking at which parts are sparse, but that would be an inaccurate guess which would have quite sever consequences, in that pieces that the disk system doesn't know whether they exist or not must be allowed to be read, and it must "fill in" the zeroes.

Another difficulty is, as you say, you accidentally set priority 0 of a file that's fully downloaded and very large. Copying all that data into the part file would be very expensive. It could even take longer than it would to re-download the file. If someone would come up with elegant and simple solutions to these challenges, we might get somewhere.

Regarding deleting the data, stock BitTorrent doesn't support advertising to peers that you no longer have a piece. There is a libtorrent extension where you can advertise that though.

>     2. can files that should be sparse, but aren't, be properly deflated? For example, a file that has 20% of its parts downloaded, but it's unchecked ("do not download"), and it has been moved, so it takes up 100% disk space. Each such file could easily free several gigabytes of space on the disk just by converting the on-disk zeros to holes (using sparse files).

They can, at least on some operating systems. But it would seem much simpler to make sure they aren't expanded to begin with, wouldn't it? Or is there some scenario you have in mind where you get sparse but fully allocated files externally somehow?

>     3. unwanted sparse files being converted to parts files, possibly in tandem with what's going on in 2.

What's the question?

**Comment by cheater** [comment]
> The main reason those "gigabytes" are there is because you move the file to a different volume, and libtorrent's copy_file() isn't sophisticated enough to preserve sparse regions, right?

Right

> So, fixing that would be an option I imagine.

Not *really*. It is one thing that needs to be fixed, but it is not a full fix. If you release this fixed tomorrow in libtorrent 3, and I upgrade my client to one that uses the new libtorrent, existing torrents would still have inflated files.

> The disk subsystem doesn't track which parts of files are sparse and not

It's not even about parts or not parts. You can make the file sparse even if a file part has been downloaded, but if that file part contained all zeros. See the "sparsify" repository on my GH: https://github.com/cheater/sparsify

Ultimately when sparsifying a file it doesn't matter whether the file just has a lot of zeros there, or if the part hasn't been downloaded. Zeros are zeros.

> Another difficulty is, as you say, you accidentally set priority 0 of a file that's fully downloaded and very large. Copying all that data into the part file would be very expensive. It could even take longer than it would to re-download the file. If someone would come up with elegant and simple solutions to these challenges, we might get somewhere.

When a file which has been downloaded partly gets set to priority zero, a client should display a dialog asking what to do with already downloaded data:
1) delete the data
2) delete all (if multiple files got set to 0 at the same time)
3) move data to parts file
4) move all data to parts file
5) make file sparse
6) make all sparse
7) keep as is
8) keep all as is

> Regarding deleting the data, stock BitTorrent doesn't support advertising to peers that you no longer have a piece.

Why is that an issue? If people request data from me, the fallback could be to not send it to them... but who knows, maybe they're running some sort of "unresponsive node" pruning algorithm that'll kick me off their swarm. Could my client re-identify in the swarm?

> But it would seem much simpler to make sure they aren't expanded to begin with, wouldn't it? 
As above - existing torrents will not be "fixed" if I upgrade to new libtorrent.

> > unwanted sparse files being converted to parts files, possibly in tandem with what's going on in 2.
> What's the question?

As you mentioned before, this is something that has not been implemented, but ideally would be.

**Comment by arvidn** [comment]
> It's not even about parts or not parts.

My paragraph that you quoted was in response to question (1), not (2). It was not about turning an allocated file into a sparse file. I was talking about moving data from a deselected file into the partfile.

> When a file which has been downloaded partly gets set to priority zero, a client should display a dialog asking what to do with already downloaded data:

At least that is outside the scope of libtorrent.

> Why is that an issue?

Because the protocol doesn't support it. There is no way to tell a client that you can't send it the data. You can disconnect and when you reconnect say you don't have the piece.

> As above - existing torrents will not be "fixed" if I upgrade to new libtorrent.

Sure, but you can use your sparsify tool as a one-off conversion.

> As you mentioned before, this is something that has not been implemented, but ideally would be.

Right, the challenge I mentioned under (1) is that either all zeroes are assumed to be sparse regions and not moved into the partfile. The negative consequence of that is that it must be acceptable to read pieces that don't exist (because maybe some zeroes had actually been written as part of a valid piece). Alternatively, the disk subsystem also tracks which pieces have been written to know which pieces to move into the partfile. The bittorrent engine already tracks this, so it ends up being double accounting. The interface to the disk subsystem could also be made more complex to have this information passed along. Neither is a particularly elegant or simple solution.

**Comment by cheater** [comment]
Understandably making the UI show up is not part of what libtorrent does. However, what happens behind this is in libtorrent's ballpark, to an important extent. Would you be happy to support such functionality?

I can't use sparsify because it only works on Linux and it's only experimental. I wouldn't use it on 100s of terabytes of my own data and I would not suggest anyone does. I also don't know how to testing it for doing the right thing. On the other hand libtorrent has additional information. To find files that need to be deflated, libtorrent can check size of file on disk vs size of file according to parts completion. It can then use checksums for checking if the deflation operation broke the file, finally moving the new file into the old file's place once things are good. I could do the latter part with a script using a theoretical windows port of sparsify (I haven't got the slightest clue how to program in Windows), but I couldn't do the former part. On top of that it would only fix the existing issue for one user, rather than everyone... Do you think you would be happy to provide that operation in libtorrent?

Regarding using sparse files to store parts data,

> Neither is a particularly elegant or simple solution.

Which one do you hate the least...?

**Comment by arvidn** [comment]
> I can't use sparsify because it only works on Linux and it's only experimental. I wouldn't use it on 100s of terabytes of my own data and I would not suggest anyone does. I also don't know how to testing it for doing the right thing.

It sounds like you're asking me to build this tool and incorporate it into libtorrent. I definitely don't think a tool like this belongs in the scope of libtorrent, and I'm also not super interested in building this tool. I'm certain this has been done many times already, for all important platforms.

> On the other hand libtorrent has additional information.

I don't think you need additional information. zeros in the file data are treated the same as a sparse region by filesystems.

**Comment by cheater** [comment]
> It sounds like you're asking me to build this tool and incorporate it into libtorrent. I definitely don't think a tool like this belongs in the scope of libtorrent, and I'm also not super interested in building this tool.

Sure, you could view it like "do I, as the developer, want to do this work, which I kinda don't want to?"

But you could also view it from the point of view of the user: "I have a file that currently is 3 GB on my drive, but after I move it, it's 36 GB." From the point of view of the user, this is an issue that is created by the software which moves that data around. In this case that's libtorrent. I think this is the more relevant point of view for this situation.

**Comment by cheater** [comment]
Bear in mind that for example the user moving this file, which is currently 3 GB, might have only 4 GB free space on the target drive. The original sparse version would fit in those 4 GB of space. The 36 GB non-sparse version would not fit. In this case, libtorrent fails to move a file for no good reason.

**Comment by arvidn** [comment]
> Bear in mind that for example the user moving this file, which is currently 3 GB, might have only 4 GB free space on the target drive.

preserving sparse regions when copying files has been fixed. That's not what I was talking about.

> The original sparse version would fit in those 4 GB of space. The 36 GB non-sparse version would not fit. In this case, libtorrent fails to move a file for no good reason. 

Does it fail for a bad reason? if so, what is it?

> So even if a reliable file existed that can make a non-sparse file sparse, it would still not be a replacement for having this feature inside of libtorrent, because the non-sparse file wouldn't fit on the target file system in the first place.

I believe you meant "tool" when you said "file". You requested that libtorrent should be able to sparsify files that had been copied with an earlier version, where sparse regions were not preserved. My previous post was arguing against that. I even quoted you to avoid any confusion about what I was responding to.

> Sure, you could view it like "do I, as the developer, want to do this work, which I kinda don't want to?"
> But you could also view it from the point of view of the user: "I have a file that currently is 3 GB on my drive, but after I move it, it's 36 GB." From the point of view of the user, this is an issue that is created by the software which moves that data around. In this case that's libtorrent. I think this is the more relevant point of view for this situation.

It sounds like you're still talking about preserving sparse regions when copying/moving files here too. I thought we had closed that topic because it has been implemented already.

**Comment by arvidn** [comment]
> How do you preserve sparse regions? Do you just create a new file and skip writing all 0's? Maybe that can be leveraged in this situation.

Yes, essentially. The main challenge is to query the source file for sparse regions. I support MacOS, Linux and Windows for this (and maybe Solaris)

**Comment by cheater** [comment]
Why not skip the query and put sparse regions any time there's a run of 0s larger than one ssd block (usually 512 KB)?

**Comment by arvidn** [comment]
> Why not skip the query and put sparse regions any time there's a run of 0s larger than one ssd block (usually 512 KB)?

Your example of a 36 GB file (3 GB on disk) would require reading 33 GB of zeroes from the filesystem. That's very expensive.

**Comment by cheater** [comment]
That's true. But it seems like almost the full code path is there already - in which case adding a flag to force this expensive way of doing things might be a solution to the problems presented here.

**Comment by cheater** [comment]
So here's coming back to "libtorrent has extra info", which is, if libtorrent knows a certain torrent block hasn't been downloaded, then it can easily be assumed to be sparse. And then, that means that no expensive memcmp is necessary.

On the other hand, if we _don't_ know these things, then a data re-check is necessary anyways, and so we're doing this expensive work already.

**Comment by pcordes** [comment]
>Yes, being able to convert files that have been expanded into new sparse files would be useful. How do you preserve sparse regions? Do you just create a new file and skip writing all 0's? Maybe that can be leveraged in this situation.

Just for the record, you can do that manually on GNU/Linux with `fallocate --dig-holes   *.mp4`  or whatever.  It searches the files for blocks of all-zeros and uses the fallocate system call to punch holes, turning those regions back into unallocated space.  This is like `cp --sparse` but without making a second copy.

The portable `posix_fallocate` system call doesn't support this, only actually allocating (making unwritten but non-sparse files).
